### ISSUE TITLE
Not detecting correct gpu platform

### ISSUE THREAD

**Issue by Chapster85** [aaronsace/MultiPoolMiner#279]
Hello,

MPM is not properly detecting the gpu platform for some of the benchmark tests.
The integrated video card (Intel) takes the gpu platform 0 and the AMD video cards are pushed to gpu platform 1.
Some of the benchmark tests assume that the gpu platform is 0 and it seems they also assume there are 6 GPUs. More specifically, 6 instances of Cryptonight-AMD (xmr-stak-amd) are launched and they just sit there, doing nothing until your script finally closes them with a "Failed".

Is there some way to specify that the gpu platform is 1 and that the number of GPUs is not 6?

I tested this with both Windows 7 and Windows 2010, completely clean install of MPM. The results were the same.

What am I doing wrong?

**Comment by aaronsace** [comment]
Ah, I've missed the platform off xmr-stak-amd; I will add it on. 

What output does this give you if you run it in PowerShell? 
`"--gpu-platform $([array]::IndexOf((Get-WmiObject -class CIM_VideoController | Select-Object -ExpandProperty AdapterCompatibility), 'Advanced Micro Devices, Inc.'))"`

**Comment by Dangerous-Mind** [comment]
$([array]::IndexOf((Get-WmiObject -class CIM_VideoController | Select-Object -ExpandProperty AdapterCompatibility), 'Advanced Micro Devices, Inc.'))

Returns 0 on my second rig. 
But I have to add --opencl-platform 2 on ethminer batch files. I haven't run the software on that rig yet.

On my first rig I had to disable the board monitor to run without that error.

**Comment by aaronsace** [comment]
@Dangerous-Mind I'll post a script for you to test in a few hours at lunch time.

**Comment by aaronsace** [comment]
Please send me the two test files output from this script: 
[test.zip](https://github.com/aaronsace/MultiPoolMiner/files/1238609/test.zip)

Also, let me know what each platform number represents on your system (like is intel platform 0 etc.).

**Comment by Dangerous-Mind** [comment]
Tested on two rigs. First with correct response as expected.
The second one pointed to the board GPU and got an error.

**Rig 1 (Platform Changed to 0 yesterday, It was 1)**
[test-wmi.txt](https://github.com/aaronsace/MultiPoolMiner/files/1239139/test-wmi.txt)


**Rig 2 (Platform should be 2)**
[15:33:58] CL Platform vendor: Intel(R) Corporation                    
[15:33:58] CL Platform name: Intel(R) OpenCL                    
[15:33:58] CL Platform version: OpenCL 2.0                     
[15:33:58] Error -1: Getting Device IDs (num)                    
[15:33:58] clDevicesNum returned error, no GPUs usable                    
[15:33:58] 0 GPU devices max detected

**Comment by aaronsace** [comment]
It says there is no GPUs on rig 2? 
Have you go the two '.txt' files from rig 2? 

> Rig 2 (Platform should be 2)
>  [15:33:58] CL Platform vendor: Intel(R) Corporation
>  [15:33:58] CL Platform name: Intel(R) OpenCL
>  [15:33:58] CL Platform version: OpenCL 2.0
>  [15:33:58] Error -1: Getting Device IDs (num)
>  [15:33:58] clDevicesNum returned error, no GPUs usable
>  [15:33:58] 0 GPU devices max detected

**Comment by Dangerous-Mind** [comment]
Sorry. Didn't see the other file.
This desktop its a mess... 

**Rig 2**
[test-wmi.txt](https://github.com/aaronsace/MultiPoolMiner/files/1239283/test-wmi.txt)
[test-sgminer.txt](https://github.com/aaronsace/MultiPoolMiner/files/1239282/test-sgminer.txt)

**Comment by aaronsace** [comment]
On rig two is AMD platform 2? Is that you are saying? 

It should be platform 0 as there is no other GPUs except AMD in your rig. 
Sgminer says you have no GPUs which is odd because how is it mining if it can't detect any GPUs. 

Can you try putting all the sgminer miner files next to the script and see if it works then?

**Comment by Dangerous-Mind** [comment]
Yes. It's platform 2.
I Did a SGMiner --gpu-platform 2 -ndev and got the GPU's

[Platform2.txt](https://github.com/aaronsace/MultiPoolMiner/files/1239472/Platform2.txt)

I don't think I have the MB graphics on on this rig has it does not show on the device manager.

**Comment by aaronsace** [comment]
Ah, getting somewhere. 

Still a bit confused as to why your cards are on 2. 
Can you send the output of platform 1 and 0 as well?

**Comment by Dangerous-Mind** [comment]
C:\Users\XXX\Desktop>SGMiner --gpu-platform 0 -ndev
[18:52:21] CL Platform vendor: Intel(R) Corporation
[18:52:21] CL Platform name: Intel(R) OpenCL
[18:52:21] CL Platform version: OpenCL 2.0
[18:52:21] Error -1: Getting Device IDs (num)
[18:52:21] clDevicesNum returned error, no GPUs usable
[18:52:21] 0 GPU devices max detected

C:\Users\XXX\Desktop>SGMiner --gpu-platform 1 -ndev
[18:52:25] CL Platform vendor: Intel(R) Corporation
[18:52:25] CL Platform name: Intel(R) OpenCL
[18:52:25] CL Platform version: OpenCL 2.0
[18:52:25] Error -1: Getting Device IDs (num)
[18:52:25] clDevicesNum returned error, no GPUs usable
[18:52:25] 0 GPU devices max detected

When installing the rig I had to disable the BIOS internal gfx because it didn't work with more then 4 GPU's. So the internal gfx should be disabled.
So I don't really know why there are two more platforms.

On rig 1 it was the igfx that created platform 1.

**Comment by aaronsace** [comment]
I hate sgminer; why does it not detect actual GPUs instead of utter nonsense platforms numbers? 
No idea how they expected people to work this out apart from guessing. 

There is no GPUs in those two platforms but I'm not sure as to what makes those two platforms show. Have you still got intel drivers installed?

**Comment by Dangerous-Mind** [comment]
Yes, I reinstalled windows and all the drivers when I bought this rig.
Probably I could solve this by myself. (adding platform x on the miners)
I just reported because some one will have this same issue.

That's why I didn't tried to fix it or installed the application on this rig.

Cant you try look on each platform for the GPU? If GPU found  and different to zero save platform to be used on sgminer and other miners.

**Comment by aaronsace** [comment]
Yes, now you have worked out how the platform thing works in sgminer then I could parse it. It's just a bit clunky loading sgminer multiple times and storing the information in a file somewhere.

**Comment by Dangerous-Mind** [comment]
I'm installing on my second rig and changed the \Bin\Cryptonight-AMDxx\config.txt to platform 2.
But they  keep getting rewritten by something.
How can I stop it from updating?

**Comment by aaronsace** [comment]
Everything is configured in the 'miners' folder.

**Comment by Dangerous-Mind** [comment]
Got it. My search was not searching inside some of the files on this machine...
My rig 2 is now 100% running. Thanks aaronsace

If you need some help on this issue...

**Comment by Dangerous-Mind** [comment]
Got one more surprise.

I changed the miner to platform 2 and got this error on sgminer:
_Error in hsa_operand section, at offset 3552:
Address offset exceeds variable size
LLVM ERROR:
 Brig container validation has failed in BRIGAsmPrinter.cpp_

According to this post https://github.com/nicehash/NiceHashMiner/issues/672 this issue was solved on sgminer-gm-5.5.5-8.
But appears to me that its not solved...

This happens in all of sgminers (AMD-Nicehash Lyra2z-AMD and the sgminer-gm-5.5.5-8 I downloaded on both rigs).
